{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests\n",
    "import shutil \n",
    "import numpy as np\n",
    "from pymol import cmd\n",
    "from pymol import stored\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "from Bio import SeqIO\n",
    "import subprocess\n",
    "from io import StringIO\n",
    "from os.path import isfile\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fixing '5E54' becoming '5.00E+54' issue\n",
    "def fix_PDB_ID(df):\n",
    "    for i, j in enumerate(df['PDB_ID']):\n",
    "        print (j)\n",
    "        if len(str(j))>4:\n",
    "            if '-' in j:\n",
    "                X= ''.join(j.split('-')).upper()\n",
    "                df.loc[i, 'PDB_ID']= X\n",
    "            else:\n",
    "                if '.' in j:\n",
    "                    x= str(j)\n",
    "                    x1= x.split('.')[0]\n",
    "                    x2= x.split('+')[1]\n",
    "                    X= x1+'E'+x2\n",
    "                    print (X)\n",
    "                    df.loc[i, 'PDB_ID']= X\n",
    "                else:\n",
    "                    x= str(j)\n",
    "                    x1= x.split('+')[0]\n",
    "                    x2= x.split('+')[1]\n",
    "                    X= x1+ x2\n",
    "                    print (X)\n",
    "                    df.loc[i, 'PDB_ID']= X\n",
    "                    \n",
    "    return (df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#function to extract structure file in cif format and convert as a pandas dataframe\n",
    "def extract_cif(p):\n",
    "    filename= p+'.cif'\n",
    "    url= 'https://files.rcsb.org/view/%s' %filename \n",
    "    u = requests.get(url).content\n",
    "    u1= str(u).split('#')\n",
    "    if len(u1)==1:\n",
    "        print ('STRUCTURE IS NOT AVAILABLE!')\n",
    "        return '0'\n",
    "    \n",
    "    else:\n",
    "        c_intrst= ['_atom_site.group_PDB', '_atom_site.id', '_atom_site.type_symbol',\n",
    "           '_atom_site.label_atom_id', '_atom_site.label_alt_id',\n",
    "           '_atom_site.label_comp_id', '_atom_site.label_asym_id',\n",
    "           '_atom_site.label_entity_id', '_atom_site.label_seq_id',\n",
    "           '_atom_site.pdbx_PDB_ins_code', '_atom_site.Cartn_x',\n",
    "           '_atom_site.Cartn_y', '_atom_site.Cartn_z', '_atom_site.occupancy',\n",
    "           '_atom_site.B_iso_or_equiv', '_atom_site.pdbx_formal_charge',\n",
    "           '_atom_site.auth_seq_id', '_atom_site.auth_comp_id',\n",
    "           '_atom_site.auth_asym_id', '_atom_site.auth_atom_id',\n",
    "           '_atom_site.pdbx_PDB_model_num', '###']\n",
    "        u2=[]\n",
    "        for i, j in enumerate(u1):\n",
    "            if c_intrst[0] in j:\n",
    "                u2.append(j)\n",
    "\n",
    "        if len(u2)==2:\n",
    "            if len(u2[0])> len(u2[1]):\n",
    "                u3= u2[0].replace('      ', '$')\n",
    "            else:\n",
    "                u3= u2[1].replace('      ', '$')\n",
    "        else:\n",
    "            u3= u2[0].replace('      ', '$')\n",
    "        u4= u3.replace('     ', '$')\n",
    "        u5= u4.replace('    ', '$')\n",
    "        u6= u5.replace('   ', '$')\n",
    "        u7= u6.replace('  ', '$')\n",
    "        u8= u7.replace(' ', '$')\n",
    "        u9= u8.replace('$$', '$')\n",
    "        u10= u9.replace('\\\\\\\\', '').split('\\\\n')\n",
    "        df = pd.DataFrame({'all_col':u10[23:]})\n",
    "\n",
    "        df[c_intrst] = df['all_col'].str.split('$', expand=True)\n",
    "        df= df.drop(['all_col', '###'], axis=1)\n",
    "\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#function to clip residues of interest and store as cif file\n",
    "def clip(p, c, s, l, n, h, m, F):\n",
    "    #p= pdb_ID, entries can be case-insensitive\n",
    "    #c= chain ID\n",
    "    #s= seg ID, this is an optional value, if you do not know the seg ID of your residues of interest, just put ''\n",
    "    #l= list of index for the residue of interest, items in this list should be strings, not integer\n",
    "    #if you want all the residues for your chain of interest, just put '' for l\n",
    "    #n= incase you have icode in the residue index for the residue of interest,\n",
    "    #instead of providing all the residue indexes as a list in l, just give the index of the starting index\n",
    "    #as a single string for l, and for n give the list of two numbers, first number will indicate how many residues \n",
    "    #before the residue of index you want, the second number will indicate how many residues after the residue of\n",
    "    #inerest you want. If the residue of interest does not contain, icode, and/or you know all the residue indexes \n",
    "    #for the residues you want, just put '' for n\n",
    "    #for instance, l= '47A', and n= [2, 7], means I want 2 residues before the residue at '47A' \n",
    "    # and 7 residue followed by residue at '47A'\n",
    "    #h= 'N' or 'Y', if 'N', that means no hydrogen coordinates should be included considering hydrogen coordinates are provided\n",
    "    #m= 'N' or 'Y', if 'N', that means no modified or non-RNA residues should be included \n",
    "    #if 'Y', that means no hydrogen coordinates should be removed if any hydrogen coordinates are provided\n",
    "    #F= desired output filename (without the .cif extension)\n",
    "    P= p.upper()\n",
    "    \n",
    "    #converting the cif file into dataframe\n",
    "    p1= extract_cif(P)\n",
    "    if len(p1)>1:\n",
    "        #applying chain ID and seg ID filter \n",
    "        if len(s) ==0:\n",
    "            p2= p1[(p1['_atom_site.auth_asym_id']==c)] \n",
    "            p2.index= np.arange(0, len(p2))\n",
    "        else:\n",
    "            p2= p1[(p1['_atom_site.auth_asym_id']==c)& (p1['_atom_site.label_asym_id']==s)] \n",
    "            p2.index= np.arange(0, len(p2))\n",
    "    \n",
    "        #converting index column values to integer\n",
    "        #p2['_atom_site.auth_seq_id'] = pd.to_numeric(p2['_atom_site.auth_seq_id'])\n",
    "        if isinstance(l, list):\n",
    "            l= [str(item) for item in l]\n",
    "            print (l)\n",
    "            p3= p2[p2['_atom_site.auth_seq_id'].isin(l)]\n",
    "            p3.index= np.arange(0, len(p3))\n",
    "        elif len(l)>0:\n",
    "            p2['index_icode']= p2['_atom_site.auth_seq_id']+ p2['_atom_site.pdbx_PDB_ins_code']\n",
    "            #print (p2['index_icode'].to_list())\n",
    "            ext_loc= int(p2[p2['index_icode'] == l]['_atom_site.label_seq_id'].to_list()[0])\n",
    "            ext_locs= []\n",
    "            for inds in range(ext_loc-n[0], ext_loc+n[1]):\n",
    "                ext_locs.append(str(inds))\n",
    "            p3= p2[p2['_atom_site.label_seq_id'].isin(ext_locs)]\n",
    "            p3.index= np.arange(0, len(p3))\n",
    "            p3 = p3.drop(columns=['index_icode'])\n",
    "        elif len(l)==0:\n",
    "            p3= p2.copy()\n",
    "        \n",
    "        #filtering out or not hydrogen atoms\n",
    "        if h== 'N':\n",
    "            p3_h= p3[p3['_atom_site.type_symbol']!= 'H']\n",
    "            p3_h.index= np.arange(0, len(p3_h))\n",
    "        else:\n",
    "            p3_h= p3.copy()\n",
    "\n",
    "        if m== 'N':\n",
    "            p3_m = p3_h[p3_h['_atom_site.auth_comp_id'].str.len() < 2]\n",
    "            p3_m.index= np.arange(0, len(p3_m))\n",
    "        else:\n",
    "            p3_m= p3_h.copy()\n",
    "    \n",
    "        #cleaning name for the sugar atoms\n",
    "        for i, j in enumerate(p3_m['_atom_site.label_atom_id']):\n",
    "            if \"\\\\'\" in j:\n",
    "                #print (j)\n",
    "                #print (j)\n",
    "                x= (j.replace(\"\\\\'\", \"'\"))\n",
    "                y= (x.replace('\"', ''))\n",
    "                p3_m.loc[i, '_atom_site.label_atom_id'] = y\n",
    "                p3_m.loc[i, '_atom_site.auth_atom_id'] = y\n",
    "        \n",
    "        #renumber the residue index starting from 1\n",
    "        \n",
    "        print ('=============================================================')\n",
    "        init_res_ind= list(set(p3_m['_atom_site.auth_seq_id'].to_list()))\n",
    "        print (init_res_ind)\n",
    "        init_res_ind.sort()\n",
    "        \n",
    "        ind_count= 0\n",
    "        ind_dict={}\n",
    "        for ind_ind, ind_val in enumerate(init_res_ind):\n",
    "            ind_count +=1\n",
    "            ind_dict[ind_val]= str(ind_count)\n",
    "        \n",
    "        p3_m['_atom_site.auth_seq_id'] = p3_m['_atom_site.auth_seq_id'].replace(ind_dict)\n",
    "        final_res_ind= list(set(p3_m['_atom_site.auth_seq_id'].to_list()))\n",
    "        print (final_res_ind)\n",
    "        print ('=============================================================')\n",
    "        \n",
    "        #Creating dataframe for the column name\n",
    "        c_intrst_1= ['_atom_site.group_PDB', '_atom_site.id', '_atom_site.type_symbol',\n",
    "               '_atom_site.label_atom_id', '_atom_site.label_alt_id',\n",
    "               '_atom_site.label_comp_id', '_atom_site.label_asym_id',\n",
    "               '_atom_site.label_entity_id', '_atom_site.label_seq_id',\n",
    "               '_atom_site.pdbx_PDB_ins_code', '_atom_site.Cartn_x',\n",
    "               '_atom_site.Cartn_y', '_atom_site.Cartn_z', '_atom_site.occupancy',\n",
    "               '_atom_site.B_iso_or_equiv', '_atom_site.pdbx_formal_charge',\n",
    "               '_atom_site.auth_seq_id', '_atom_site.auth_comp_id',\n",
    "               '_atom_site.auth_asym_id', '_atom_site.auth_atom_id',\n",
    "               '_atom_site.pdbx_PDB_model_num']\n",
    "        c_intrst_2= ['data_', '#','loop_', '_atom_site.group_PDB', '_atom_site.id', '_atom_site.type_symbol',\n",
    "               '_atom_site.label_atom_id', '_atom_site.label_alt_id',\n",
    "               '_atom_site.label_comp_id', '_atom_site.label_asym_id',\n",
    "               '_atom_site.label_entity_id', '_atom_site.label_seq_id',\n",
    "               '_atom_site.pdbx_PDB_ins_code', '_atom_site.Cartn_x',\n",
    "               '_atom_site.Cartn_y', '_atom_site.Cartn_z', '_atom_site.occupancy',\n",
    "               '_atom_site.B_iso_or_equiv', '_atom_site.pdbx_formal_charge',\n",
    "               '_atom_site.auth_seq_id', '_atom_site.auth_comp_id',\n",
    "               '_atom_site.auth_asym_id', '_atom_site.auth_atom_id',\n",
    "               '_atom_site.pdbx_PDB_model_num']\n",
    "    \n",
    "        c_intrst_2[0]= 'data_'+P\n",
    "    \n",
    "        p4= pd.DataFrame(columns= c_intrst_1)\n",
    "        p4['_atom_site.group_PDB']= c_intrst_2 #this p3 dataframe above p2 will be joined in p4. p3 will help pymol to visualize the file\n",
    "    \n",
    "        p5 = pd.concat([p4, p3_m], axis=0)\n",
    "        \n",
    "        \n",
    "        \n",
    "    \n",
    "        p5.to_csv(F+'.cif', header= False, sep=' ', index= False) #unmute this\n",
    "        return p3_m\n",
    "        #print ('WE COMPLETED THIS FARRRRRRRR')\n",
    "        #p5.to_csv(P+'_clipped.cif', header= False, sep=' ', index= False)\n",
    "    else:\n",
    "        print ('STRUCTURE IS NOT AVAILABLE!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_cls_str(df, tar_dir):\n",
    "    \n",
    "    #tar_dir= '/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/test'\n",
    "    home= os.getcwd()\n",
    "    \n",
    "    os.chdir(tar_dir)\n",
    "    df = df.dropna(subset=['cluster'])\n",
    "    print (df.shape)\n",
    "    cls_names= [cls_name for cls_name in list(set(df['cluster'].to_list())) if cls_name != 'unclustered']\n",
    "    cls_names.sort()\n",
    "    #cls_names= cls_name_i.sort()\n",
    "    print (cls_names)\n",
    "        \n",
    "    df['bp_ID']= df['bp_ID'].str.replace('-', '_', regex=False)\n",
    "    \n",
    "    for i, j in enumerate(cls_names):\n",
    "        if j== 'C1':\n",
    "            fln= 3\n",
    "            \n",
    "        elif j== 'C2':\n",
    "            fln= 4\n",
    "            \n",
    "        elif j== 'C3':\n",
    "            fln= 9\n",
    "            \n",
    "        elif j== 'C4':\n",
    "            fln= 6\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        cls_path = os.path.join(tar_dir, j)\n",
    "        os.makedirs(cls_path, exist_ok=True)\n",
    "        \n",
    "        os.chdir(cls_path)\n",
    "        \n",
    "        df_c= df[df['cluster']== j]\n",
    "        df_c.index= np.arange(0, len(df_c))\n",
    "        for k, l in enumerate(df_c['bp_ID']):\n",
    "            res1_index= int(df_c['res_index_res1'][k])\n",
    "            res2_index= int(df_c['res_index_res2'][k])\n",
    "            p= df_c['PDB_ID'][k]\n",
    "            c= df_c['chain_ID'][k]\n",
    "            s= df_c['seg_ID'][k]\n",
    "            resn_list_motif= [k for k in range(res1_index-fln, res2_index+fln+1)]\n",
    "            resn_list_motif.sort()\n",
    "            filename= j+'_'+l\n",
    "            cls_str= clip(p, c, s, resn_list_motif, '', 'N', 'Y', filename)\n",
    "            \n",
    "        os.chdir(home)\n",
    "\n",
    "    os.chdir(home)\n",
    "    #return all_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/manuscript_things/figshare/scripts/plot_scripts/fig_7')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the final output data for the shifted wobble\n",
    "sw_df= pd.read_csv('../../../results/all_shifted_wobble_with_non_WCF_interaction.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/manuscript_things/figshare/scripts/plot_scripts/fig_7\n"
     ]
    }
   ],
   "source": [
    "print (os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# clipping motifs containing clusterd wobbles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(373, 93)\n",
      "['C1', 'C2', 'C3', 'C4']\n",
      "['1080', '1081', '1082', '1083', '1084', '1085', '1086', '1087', '1088', '1089', '1090', '1091', '1092', '1093', '1094', '1095', '1096', '1097', '1098', '1099']\n",
      "=============================================================\n",
      "['1099', '1089', '1082', '1087', '1097', '1091', '1083', '1081', '1094', '1093', '1086', '1088', '1090', '1092', '1084', '1085', '1080', '1096', '1098', '1095']\n",
      "['4', '6', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['1083', '1084', '1085', '1086', '1087', '1088', '1089', '1090', '1091', '1092', '1093', '1094', '1095', '1096', '1097', '1098', '1099', '1100', '1101', '1102']\n",
      "=============================================================\n",
      "['1099', '1089', '1087', '1097', '1091', '1083', '1094', '1093', '1086', '1088', '1090', '1092', '1084', '1085', '1100', '1102', '1096', '1098', '1095', '1101']\n",
      "['4', '6', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['1077', '1078', '1079', '1080', '1081', '1082', '1083', '1084', '1085', '1086', '1087', '1088', '1089', '1090', '1091', '1092', '1093', '1094', '1095', '1096']\n",
      "=============================================================\n",
      "['1089', '1082', '1087', '1091', '1083', '1081', '1094', '1093', '1077', '1086', '1088', '1078', '1079', '1090', '1092', '1084', '1085', '1080', '1096', '1095']\n",
      "['4', '6', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['1094', '1095', '1096', '1097', '1098', '1099', '1100', '1101', '1102', '1103', '1104', '1105', '1106', '1107', '1108', '1109', '1110', '1111', '1112', '1113']\n",
      "=============================================================\n",
      "['1099', '1111', '1097', '1107', '1106', '1094', '1105', '1113', '1112', '1108', '1109', '1103', '1100', '1102', '1096', '1098', '1095', '1104', '1110', '1101']\n",
      "['4', '6', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['1083', '1084', '1085', '1086', '1087', '1088', '1089', '1090', '1091', '1092', '1093', '1094', '1095', '1096', '1097', '1098', '1099', '1100', '1101', '1102']\n",
      "=============================================================\n",
      "['1099', '1089', '1087', '1097', '1091', '1083', '1094', '1093', '1086', '1088', '1090', '1092', '1084', '1085', '1100', '1102', '1096', '1098', '1095', '1101']\n",
      "['4', '6', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['670', '671', '672', '673', '674', '675', '676', '677', '678', '679', '680', '681', '682', '683', '684', '685', '686', '687', '688', '689', '690', '691', '692', '693', '694', '695', '696', '697', '698', '699', '700', '701', '702', '703', '704', '705', '706', '707', '708', '709', '710', '711', '712', '713', '714']\n",
      "=============================================================\n",
      "['670', '700', '673', '703', '706', '707', '710', '677', '686', '679', '712', '684', '683', '689', '709', '685', '681', '708', '691', '675', '711', '702', '672', '692', '678', '699', '701', '688', '690', '698', '704', '694', '695', '671', '714', '687', '713', '696', '682', '674', '693', '705', '697', '680', '676']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['682', '683', '684', '685', '686', '687', '688', '689', '690', '691', '692', '693', '694', '695', '696', '697', '698', '699', '700', '701', '702', '703', '704', '705', '706', '707', '708', '709', '710', '711', '712', '713', '714', '715', '716', '717', '718', '719', '720', '721', '722', '723', '724', '725', '726']\n",
      "=============================================================\n",
      "['700', '703', '706', '707', '710', '686', '712', '684', '724', '683', '689', '709', '685', '726', '718', '708', '691', '711', '716', '702', '715', '692', '699', '701', '688', '690', '698', '704', '694', '695', '714', '687', '713', '719', '696', '723', '682', '721', '722', '717', '725', '693', '705', '720', '697']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['673', '674', '675', '676', '677', '678', '679', '680', '681', '682', '683', '684', '685', '686', '687', '688', '689', '690', '691', '692', '693', '694', '695', '696', '697', '698', '699', '700', '701', '702', '703', '704', '705', '706', '707', '708', '709', '710', '711', '712', '713', '714', '715', '716', '717']\n",
      "=============================================================\n",
      "['700', '703', '673', '706', '707', '710', '677', '686', '679', '712', '684', '683', '689', '709', '685', '681', '708', '691', '675', '711', '716', '702', '715', '692', '699', '678', '701', '688', '690', '698', '704', '694', '695', '714', '687', '713', '696', '682', '674', '717', '693', '705', '697', '680', '676']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['683', '684', '685', '686', '687', '688', '689', '690', '691', '692', '693', '694', '695', '696', '697', '698', '699', '700', '701', '702', '703', '704', '705', '706', '707', '708', '709', '710', '711', '712', '713', '714', '715', '716', '717', '718', '719', '720', '721', '722', '723', '724', '725', '726', '727']\n",
      "=============================================================\n",
      "['700', '703', '706', '707', '710', '686', '712', '684', '724', '683', '689', '709', '685', '726', '718', '708', '691', '711', '716', '702', '715', '692', '699', '701', '688', '690', '698', '704', '694', '695', '714', '687', '713', '719', '696', '723', '721', '722', '717', '725', '693', '705', '720', '697', '727']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['664', '665', '666', '667', '668', '669', '670', '671', '672', '673', '674', '675', '676', '677', '678', '679', '680', '681', '682', '683', '684', '685', '686', '687', '688', '689', '690', '691', '692', '693', '694', '695', '696', '697', '698', '699', '700', '701', '702', '703', '704', '705', '706', '707', '708']\n",
      "=============================================================\n",
      "['670', '700', '673', '703', '706', '707', '677', '686', '679', '684', '683', '689', '685', '681', '708', '691', '675', '666', '702', '672', '692', '678', '699', '701', '665', '688', '690', '698', '694', '695', '704', '669', '671', '664', '687', '696', '682', '668', '674', '667', '693', '705', '697', '680', '676']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['667', '668', '669', '670', '671', '672', '673', '674', '675', '676', '677', '678', '679', '680', '681', '682', '683', '684', '685', '686', '687', '688', '689', '690', '691', '692', '693', '694', '695', '696', '697', '698', '699', '700', '701', '702', '703', '704', '705', '706', '707', '708', '709', '710', '711']\n",
      "=============================================================\n",
      "['670', '700', '673', '703', '706', '707', '710', '677', '686', '679', '684', '683', '689', '709', '685', '681', '708', '691', '675', '711', '702', '672', '692', '678', '699', '701', '688', '690', '698', '704', '694', '695', '671', '669', '687', '696', '682', '668', '674', '667', '693', '705', '697', '680', '676']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['681', '682', '683', '684', '685', '686', '687', '688', '689', '690', '691', '692', '693', '694', '695', '696', '697', '698', '699', '700', '701', '702', '703', '704', '705', '706', '707', '708', '709', '710', '711', '712', '713', '714', '715', '716', '717', '718', '719', '720', '721', '722', '723', '724', '725']\n",
      "=============================================================\n",
      "['700', '703', '706', '707', '710', '686', '712', '684', '724', '683', '689', '709', '685', '681', '718', '708', '691', '711', '716', '702', '715', '692', '699', '701', '688', '690', '698', '704', '694', '695', '714', '687', '713', '719', '696', '723', '682', '721', '722', '717', '725', '693', '705', '720', '697']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['673', '674', '675', '676', '677', '678', '679', '680', '681', '682', '683', '684', '685', '686', '687', '688', '689', '690', '691', '692', '693', '694', '695', '696', '697', '698', '699', '700', '701', '702', '703', '704', '705', '706', '707', '708', '709', '710', '711', '712', '713', '714', '715', '716', '717']\n",
      "=============================================================\n",
      "['700', '703', '673', '706', '707', '710', '677', '686', '679', '712', '684', '683', '689', '709', '685', '681', '708', '691', '675', '711', '716', '702', '715', '692', '699', '678', '701', '688', '690', '698', '704', '694', '695', '714', '687', '713', '696', '682', '674', '717', '693', '705', '697', '680', '676']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['552', '553', '554', '555', '556', '557', '558', '559', '560', '561', '562', '563', '564', '565', '566', '567', '568', '569', '570', '571', '572', '573', '574', '575', '576', '577', '578', '579', '580', '581', '582', '583', '584', '585', '586', '587', '588', '589', '590', '591', '592', '593', '594', '595', '596']\n",
      "=============================================================\n",
      "['576', '557', '585', '564', '596', '567', '566', '591', '556', '584', '590', '586', '573', '568', '588', '589', '581', '575', '580', '583', '563', '579', '560', '574', '562', '577', '558', '570', '592', '571', '555', '554', '578', '553', '559', '594', '561', '593', '572', '565', '587', '552', '582', '569', '595']\n",
      "['26', '37', '4', '36', '42', '22', '43', '6', '45', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '40', '24', '30', '41', '13', '31', '44', '33', '39', '35', '18', '7', '11', '38', '20', '28', '32', '9', '23', '14', '34', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['462', '463', '464', '465', '466', '467', '468', '469', '470', '471', '472', '473', '474', '475', '476', '477', '478', '479', '480', '481', '482', '483', '484', '485', '486', '487', '488', '489', '490']\n",
      "=============================================================\n",
      "['479', '490', '478', '477', '473', '468', '471', '483', '487', '465', '485', '467', '463', '489', '464', '476', '484', '474', '472', '475', '470', '466', '462', '480', '488', '482', '469', '481', '486']\n",
      "['26', '4', '22', '6', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '24', '13', '18', '7', '11', '20', '28', '9', '23', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['461', '462', '463', '464', '465', '466', '467', '468', '469', '470', '471', '472', '473', '474', '475', '476', '477', '478', '479', '480', '481', '482', '483', '484', '485', '486', '487', '488', '489']\n",
      "=============================================================\n",
      "['479', '478', '477', '473', '468', '471', '483', '487', '465', '485', '467', '463', '489', '461', '464', '476', '484', '474', '472', '475', '470', '466', '462', '480', '488', '482', '469', '481', '486']\n",
      "['26', '4', '22', '6', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '24', '13', '18', '7', '11', '20', '28', '9', '23', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['385', '386', '387', '388', '389', '390', '391', '392', '393', '394', '395', '396', '397', '398', '399', '400', '401', '402', '403', '404', '405', '406', '407', '408', '409', '410', '411', '412', '413']\n",
      "=============================================================\n",
      "['396', '398', '406', '409', '395', '391', '404', '399', '413', '407', '403', '390', '408', '387', '388', '400', '410', '397', '412', '386', '389', '392', '385', '401', '402', '405', '394', '393', '411']\n",
      "['26', '4', '22', '6', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '24', '13', '18', '7', '11', '20', '28', '9', '23', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['419', '420', '421', '422', '423', '424', '425', '426', '427', '428', '429', '430', '431', '432', '433', '434', '435', '436', '437', '438', '439', '440', '441', '442', '443', '444', '445', '446', '447']\n",
      "=============================================================\n",
      "['432', '440', '443', '447', '445', '425', '428', '430', '421', '439', '426', '420', '436', '429', '431', '435', '434', '438', '427', '446', '423', '437', '422', '442', '424', '444', '441', '419', '433']\n",
      "['26', '4', '22', '6', '21', '2', '5', '17', '15', '25', '19', '29', '10', '27', '24', '13', '18', '7', '11', '20', '28', '9', '23', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['2312', '2313', '2314', '2315', '2316', '2317', '2318', '2319', '2320', '2321', '2322', '2323', '2324', '2325', '2326', '2327', '2328', '2329', '2330', '2331', '2332']\n",
      "=============================================================\n",
      "['2316', '2318', '2328', '2315', '2327', '2321', '2330', '2319', '2324', '2331', '2332', '2329', '2325', '2323', '2320', '2314', '2326', '2313', '2322', '2317', '2312']\n",
      "['4', '6', '21', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['2298', '2299', '2300', '2301', '2302', '2303', '2304', '2305', '2306', '2307', '2308', '2309', '2310', '2311', '2312', '2313', '2314', '2315', '2316', '2317', '2318']\n",
      "=============================================================\n",
      "['2316', '2318', '2315', '2309', '2310', '2300', '2305', '2299', '2308', '2301', '2302', '2306', '2314', '2304', '2311', '2307', '2303', '2298', '2313', '2317', '2312']\n",
      "['4', '6', '21', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['2331', '2332', '2333', '2334', '2335', '2336', '2337', '2338', '2339', '2340', '2341', '2342', '2343', '2344', '2345', '2346', '2347', '2348', '2349', '2350', '2351']\n",
      "=============================================================\n",
      "['2351', '2342', '2344', '2341', '2350', '2339', '2331', '2332', '2335', '2343', '2333', '2348', '2346', '2336', '2345', '2334', '2349', '2340', '2347', '2337', '2338']\n",
      "['4', '6', '21', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['2536', '2537', '2538', '2539', '2540', '2541', '2542', '2543', '2544', '2545', '2546', '2547', '2548', '2549', '2550', '2551', '2552', '2553', '2554', '2555', '2556']\n",
      "=============================================================\n",
      "['2545', '2550', '2546', '2537', '2540', '2556', '2536', '2543', '2551', '2538', '2549', '2555', '2539', '2542', '2554', '2541', '2544', '2547', '2552', '2553', '2548']\n",
      "['4', '6', '21', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n",
      "['2285', '2286', '2287', '2288', '2289', '2290', '2291', '2292', '2293', '2294', '2295', '2296', '2297', '2298', '2299', '2300', '2301', '2302', '2303', '2304', '2305']\n",
      "=============================================================\n",
      "['2290', '2297', '2293', '2295', '2300', '2296', '2289', '2305', '2288', '2299', '2301', '2302', '2287', '2304', '2291', '2286', '2303', '2285', '2298', '2294', '2292']\n",
      "['4', '6', '21', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['2298', '2299', '2300', '2301', '2302', '2303', '2304', '2305', '2306', '2307', '2308', '2309', '2310', '2311', '2312', '2313', '2314', '2315', '2316', '2317', '2318']\n",
      "=============================================================\n",
      "['2316', '2318', '2315', '2309', '2310', '2300', '2305', '2299', '2308', '2301', '2302', '2306', '2314', '2304', '2311', '2307', '2303', '2298', '2313', '2317', '2312']\n",
      "['4', '6', '21', '2', '5', '17', '15', '19', '10', '13', '18', '7', '11', '20', '9', '14', '1', '12', '3', '16', '8']\n",
      "=============================================================\n"
     ]
    }
   ],
   "source": [
    "generate_cls_str(sw_df, os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# characterizing, dbn, fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "home= os.getcwd()\n",
    "\n",
    "C1_dir= '/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/C1'\n",
    "C2_dir= '/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/C2'\n",
    "C3_dir= '/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/C3'\n",
    "C4_dir= '/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/C4'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chr_dbn_fa(wdir, fl_length):\n",
    "    \n",
    "    dir= os.getcwd()\n",
    "    os.chdir(wdir)\n",
    "    count= 0\n",
    "    all_seqs={}\n",
    "    for filename in os.listdir('.'):\n",
    "        if filename.endswith('.cif'):\n",
    "            count +=1\n",
    "            print (count)\n",
    "            #print (filename)\n",
    "            pname= filename[:len(filename)-4]+'-dssr.json'\n",
    "            #print (pname)\n",
    "            x= dir+'/'+ pname\n",
    "            if isfile(x)== True:\n",
    "                print ('ALREADY CHARACTERIZED')\n",
    "            else:\n",
    "                print ('NOT CHARACTERIZED YET')\n",
    "                print (filename)\n",
    "                #characterizing\n",
    "                os.system('./x3dna-dssr -i='+filename+ ' --json -o='+ filename[:len(filename)-4]+'-dssr.json')\n",
    "                \n",
    "                #open the json file\n",
    "                with open(filename[:len(filename)-4]+'-dssr.json', 'r') as f:\n",
    "                    data= json.loads(f.read())\n",
    "    \n",
    "                bps_df= pd.json_normalize(data, record_path =['pairs'])\n",
    "                bp_dict= {}\n",
    "                for nti, ntj in enumerate(bps_df['nt1']):\n",
    "                    nt1_r= ntj.split('.')[1][1:]\n",
    "                    nt2_r= bps_df['nt2'][nti].split('.')[1][1:]  \n",
    "                    if nt1_r in bp_dict:\n",
    "                        pass\n",
    "                    else:\n",
    "                        bp_dict[nt1_r]= nt2_r\n",
    "                \n",
    "                \n",
    "                print (bp_dict)\n",
    "                #generating dbn\n",
    "                for f in os.listdir('.'):\n",
    "                    if f=='dssr-2ndstrs.dbn':\n",
    "                        dbn_df= pd.read_csv(f)\n",
    "                        cols= dbn_df.columns[0]\n",
    "                        seq= dbn_df[cols][0]\n",
    "                        dbn_list=['.']*len(seq)\n",
    "                        \n",
    "                        dbn_list[fl_length]= '('\n",
    "                        dbn_list[len(seq)-fl_length-1]= ')'\n",
    "                        \n",
    "                        seq_loc= 0\n",
    "                        wrs= [str(fl_length+1), str(len(seq)-fl_length)] #residue indexes for the residues forming shifted wobble\n",
    "                        \n",
    "                        for x1 in bp_dict:\n",
    "                            if x1 in wrs or bp_dict[x1] in wrs:\n",
    "                                pass\n",
    "                            else:\n",
    "                                dbn_list[int(x1)-1]= '('\n",
    "                                dbn_list[int(bp_dict[x1])-1]= ')'\n",
    "                        \n",
    "                        #for x1, x2 in enumerate(dbn_list):\n",
    "                        #    if str(x1+1) in bp_dict:\n",
    "                        #        print ('--------------------')\n",
    "                        #        print (bp_dict[str(x1+1)])\n",
    "                        #        dbn_list[x1]= '('\n",
    "                        #        dbn_list[int(bp_dict[str(x1+1)])-1]= ')'\n",
    "                        #    else:\n",
    "                        #        pass\n",
    "                          \n",
    "                        #sec_str= dbn_df[cols][1]\n",
    "                        sec_str= ''.join(dbn_list)\n",
    "                    \n",
    "                        print ('We did it')\n",
    "                        print (seq)\n",
    "                        print (sec_str)\n",
    "                        print ('We are done')\n",
    "                        all_seqs[cols]= seq\n",
    "                    \n",
    "                        with open(filename[:len(filename)-4]+'.dbn', 'w') as file:\n",
    "                            file.write(dbn_df.columns[0] + '\\n')\n",
    "                            file.write(seq + '\\n')\n",
    "                            file.write(sec_str + '\\n')\n",
    "\n",
    "\n",
    "                    if f.startswith('dssr-'):\n",
    "                        os.remove(f)\n",
    "\n",
    "    #generating .fa file                    \n",
    "    seq_fname= list(set([i[:2] for i in os.listdir('.') if i.endswith('.cif')]))[0]+ '_seqs.fa'\n",
    "    with open(seq_fname, \"w\") as f:\n",
    "        for i in all_seqs:\n",
    "            f.write('>'+ i[4:] + '\\n')\n",
    "            f.write(all_seqs[i] + '\\n')\n",
    "    os.chdir(dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "NOT CHARACTERIZED YET\n",
      "C1_7M4Y_a_U1083_a_G1096.cif\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing file 'C1_7M4Y_a_U1083_a_G1096.cif'\n",
      "[i] 'C1_7M4Y_a_U1083_a_G1096.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 20\n",
      "    total number of base pairs: 6\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 4\n",
      "    total number of splayed-apart dinucleotides: 4\n",
      "                        consolidated into units: 3\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C1_7UNU_a_U1080_a_G1093.cif'\n",
      "[i] 'C1_7UNU_a_U1080_a_G1093.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 20\n",
      "    total number of base pairs: 7\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 2\n",
      "    total number of splayed-apart dinucleotides: 3\n",
      "                        consolidated into units: 2\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C1_4B3R_A_U1086_A_G1099.cif'\n",
      "[i] 'C1_4B3R_A_U1086_A_G1099.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 20\n",
      "    total number of base pairs: 7\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 3\n",
      "    total number of splayed-apart dinucleotides: 3\n",
      "                        consolidated into units: 2\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C1_7ZTA_16S1_U1086_16S1_G1099.cif'\n",
      "[i] 'C1_7ZTA_16S1_U1086_16S1_G1099.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 20\n",
      "    total number of base pairs: 6\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 4\n",
      "    total number of splayed-apart dinucleotides: 3\n",
      "                        consolidated into units: 2\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C1_6YEF_a_U1097_a_G1110.cif'\n",
      "[i] 'C1_6YEF_a_U1097_a_G1110.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 20\n",
      "    total number of base pairs: 7\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 3\n",
      "    total number of splayed-apart dinucleotides: 4\n",
      "                        consolidated into units: 3\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'2': '20', '4': '17', '5': '16', '6': '15', '7': '14', '8': '13'}\n",
      "We did it\n",
      "UGUUGGGUUAAGUCCCGCAA\n",
      ".(.(((((....)))))..)\n",
      "We are done\n",
      "2\n",
      "NOT CHARACTERIZED YET\n",
      "C1_7UNU_a_U1080_a_G1093.cif\n",
      "{'2': '20', '4': '17', '5': '16', '6': '15', '7': '14', '8': '13', '9': '11'}\n",
      "We did it\n",
      "UGUUGGGUUAAGUCCCGUAA\n",
      ".(.((((((.).)))))..)\n",
      "We are done\n",
      "3\n",
      "NOT CHARACTERIZED YET\n",
      "C1_4B3R_A_U1086_A_G1099.cif\n",
      "{'2': '20', '4': '17', '5': '16', '6': '15', '7': '14', '8': '13', '9': '11'}\n",
      "We did it\n",
      "UGUUGGGUUAAGUCCCGCAA\n",
      ".(.((((((.).)))))..)\n",
      "We are done\n",
      "4\n",
      "NOT CHARACTERIZED YET\n",
      "C1_7ZTA_16S1_U1086_16S1_G1099.cif\n",
      "{'2': '20', '4': '17', '5': '16', '6': '15', '7': '14', '8': '13'}\n",
      "We did it\n",
      "UGUUGGGUUAAGUCCCGCAA\n",
      ".(.(((((....)))))..)\n",
      "We are done\n",
      "5\n",
      "NOT CHARACTERIZED YET\n",
      "C1_6YEF_a_U1097_a_G1110.cif\n",
      "{'2': '20', '4': '17', '5': '16', '6': '15', '7': '14', '8': '13', '9': '11'}\n",
      "We did it\n",
      "UGUUGGGUUAAGUCCCGCAA\n",
      ".(.((((((.).)))))..)\n",
      "We are done\n",
      "1\n",
      "NOT CHARACTERIZED YET\n",
      "C2_4V8I_AA_U677_AA_G713.cif\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C2_4V8I_AA_U677_AA_G713.cif'\n",
      "[i] 'C2_4V8I_AA_U677_AA_G713.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 20\n",
      "    total number of multiplets: 1\n",
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of atom-base capping interactions: 2\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of eXtended A-minor (type X) motifs: 1\n",
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C2_7QEP_3_U556_3_G592.cif'\n",
      "[i] 'C2_7QEP_3_U556_3_G592.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 20\n",
      "    total number of multiplets: 2\n",
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of A-minor (types I and II) motifs: 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '28', '16': '27', '17': '26', '18': '25', '19': '24'}\n",
      "We did it\n",
      "GGAAUUCCCGGAGUAGCGGUGAAAUGCGCAGAUACCGGGAGGAAC\n",
      "(((((((((((((((((((....)))))...))))))))))))))\n",
      "We are done\n",
      "2\n",
      "NOT CHARACTERIZED YET\n",
      "C2_7QEP_3_U556_3_G592.cif\n",
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '28', '16': '27', '17': '26', '18': '25'}\n",
      "We did it\n",
      "UGGAUAGUUGGGCGAGAGGUGAAAUGCGAAGACCCUGACUGGACG\n",
      "((((((((((((((((((......))))...))))))))))))))\n",
      "We are done\n",
      "3\n",
      "NOT CHARACTERIZED YET\n",
      "C2_6YEF_a_U685_a_G721.cif\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C2_6YEF_a_U685_a_G721.cif'\n",
      "[i] 'C2_6YEF_a_U685_a_G721.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 21\n",
      "    total number of multiplets: 2\n",
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of A-minor (types I and II) motifs: 1\n",
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '28', '16': '27', '17': '26', '18': '25', '19': '24'}\n",
      "We did it\n",
      "GGAAUUCCAUGUGUAGCGGUGAAAUGCGCAGAGAUAUGGAGGAAC\n",
      "(((((((((((((((((((....)))))...))))))))))))))\n",
      "We are done\n",
      "4\n",
      "NOT CHARACTERIZED YET\n",
      "C2_8B0X_A_U677_A_G713.cif\n",
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '28', '16': '27', '17': '26', '18': '25', '19': '24'}\n",
      "We did it\n",
      "AGAAUUCCAGGUGUAGCGGUGAAAUGCGUAGAGAUCUGGAGGAAU\n",
      "(((((((((((((((((((....)))))...))))))))))))))\n",
      "We are done\n",
      "5\n",
      "NOT CHARACTERIZED YET\n",
      "C2_7UVZ_a_U674_a_G710.cif\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing file 'C2_8B0X_A_U677_A_G713.cif'\n",
      "[i] 'C2_8B0X_A_U677_A_G713.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 21\n",
      "    total number of multiplets: 2\n",
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of atom-base capping interactions: 2\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of A-minor (types I and II) motifs: 1\n",
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C2_7UVZ_a_U674_a_G710.cif'\n",
      "[i] 'C2_7UVZ_a_U674_a_G710.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 20\n",
      "    total number of multiplets: 1\n",
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of atom-base capping interactions: 2\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of A-minor (types I and II) motifs: 1\n",
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C2_6HA1_a_U686_a_G722.cif'\n",
      "[i] 'C2_6HA1_a_U686_a_G722.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 20\n",
      "    total number of multiplets: 1\n",
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of eXtended A-minor (type X) motifs: 1\n",
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C2_7NHN_a_U687_a_G723.cif'\n",
      "[i] 'C2_7NHN_a_U687_a_G723.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 19\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '31', '16': '27', '17': '26', '18': '25', '19': '24'}\n",
      "We did it\n",
      "AGAAUUCCAGGUGUAGCGGUGAAAUGCGUAGAGAUCUGGAGGAAU\n",
      "(((((((((((((((((((....))))...)))))))))))))))\n",
      "We are done\n",
      "6\n",
      "NOT CHARACTERIZED YET\n",
      "C2_6HA1_a_U686_a_G722.cif\n",
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '31', '16': '27', '17': '26', '18': '25', '19': '24'}\n",
      "We did it\n",
      "GGAAUUCCACGUGUAGCGGUGAAAUGCGUAGAGAUGUGGAGGAAC\n",
      "(((((((((((((((((((....))))...)))))))))))))))\n",
      "We are done\n",
      "7\n",
      "NOT CHARACTERIZED YET\n",
      "C2_7NHN_a_U687_a_G723.cif\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of eXtended A-minor (type X) motifs: 1\n",
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C2_7UNR_a_U671_a_G707.cif'\n",
      "[i] 'C2_7UNR_a_U671_a_G707.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 21\n",
      "    total number of multiplets: 1\n",
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of A-minor (types I and II) motifs: 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '31', '16': '27', '17': '26', '18': '25', '19': '24'}\n",
      "We did it\n",
      "GGAAUUCCACGUGUAGCGGUGAAAUGCGUAGAUAUGUGGAGGAAC\n",
      "(((((((((((((((((((....))))...)))))))))))))))\n",
      "We are done\n",
      "8\n",
      "NOT CHARACTERIZED YET\n",
      "C2_7UNR_a_U671_a_G707.cif\n",
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '31', '16': '27', '17': '26', '18': '25', '19': '24', '20': '23'}\n",
      "We did it\n",
      "GGAAUUUCCUGUGUAGCGGUGAAAUGCGUAGAUAUAGGAAGGAAC\n",
      "((((((((((((((((((((..)))))...)))))))))))))))\n",
      "We are done\n",
      "9\n",
      "NOT CHARACTERIZED YET\n",
      "C2_7MT7_a_U668_a_G704.cif\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C2_7MT7_a_U668_a_G704.cif'\n",
      "[i] 'C2_7MT7_a_U668_a_G704.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 45\n",
      "    total number of base pairs: 21\n",
      "    total number of multiplets: 2\n",
      "    total number of helices: 2\n",
      "    total number of stems: 2\n",
      "    total number of isolated WC/wobble pairs: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of internal loops: 2\n",
      "    total number of A-minor (types I and II) motifs: 1\n",
      "    total number of kink turns: 1\n",
      "\n",
      "Time used: 00:00:00:00\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': '45', '2': '44', '3': '43', '4': '42', '5': '41', '6': '40', '7': '39', '8': '38', '9': '37', '10': '36', '11': '35', '12': '34', '13': '33', '14': '32', '15': '28', '16': '27', '17': '26', '18': '25', '19': '24'}\n",
      "We did it\n",
      "GGAAUUCCUGGUGUAGCGGUGGAAUGCGCAGAUAUCAGGAGGAAC\n",
      "(((((((((((((((((((....)))))...))))))))))))))\n",
      "We are done\n",
      "1\n",
      "NOT CHARACTERIZED YET\n",
      "C3_5MMI_A_G394_A_U404.cif\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing file 'C3_5MMI_A_G394_A_U404.cif'\n",
      "[i] 'C3_5MMI_A_G394_A_U404.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 29\n",
      "    total number of base pairs: 11\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C3_5NGM_AA_G428_AA_U438.cif'\n",
      "[i] 'C3_5NGM_AA_G428_AA_U438.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 29\n",
      "    total number of base pairs: 13\n",
      "    total number of multiplets: 1\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 2\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "    total number of eXtended A-minor (type X) motifs: 1\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C3_7MT2_A_G471_A_U481.cif'\n",
      "[i] 'C3_7MT2_A_G471_A_U481.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 29\n",
      "    total number of base pairs: 11\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C3_7S0S_C_G470_C_U480.cif'\n",
      "[i] 'C3_7S0S_C_G470_C_U480.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 29\n",
      "    total number of base pairs: 11\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of splayed-apart dinucleotides: 2\n",
      "                        consolidated into units: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': '29', '2': '28', '3': '27', '4': '26', '5': '25', '6': '24', '7': '23', '8': '22', '9': '21', '10': '20', '11': '19'}\n",
      "We did it\n",
      "UAGCAUGGGGCACGUGGAAUCCCGUGUGA\n",
      "(((((((((((.......)))))))))))\n",
      "We are done\n",
      "2\n",
      "NOT CHARACTERIZED YET\n",
      "C3_5NGM_AA_G428_AA_U438.cif\n",
      "{'1': '29', '2': '28', '3': '27', '4': '26', '5': '25', '6': '24', '7': '23', '8': '22', '9': '15', '10': '20', '11': '18'}\n",
      "We did it\n",
      "UACGACGGAGCACGUGAAAUUCCGUCGGA\n",
      "(((((((((((...)..).).))))))))\n",
      "We are done\n",
      "3\n",
      "NOT CHARACTERIZED YET\n",
      "C3_7MT2_A_G471_A_U481.cif\n",
      "{'1': '29', '2': '28', '3': '27', '4': '26', '5': '25', '6': '24', '7': '23', '8': '22', '9': '21', '10': '20', '11': '19'}\n",
      "We did it\n",
      "UAGCAGCGGGCCCGUGGAAUCCGCUGUGA\n",
      "(((((((((((.......)))))))))))\n",
      "We are done\n",
      "4\n",
      "NOT CHARACTERIZED YET\n",
      "C3_7S0S_C_G470_C_U480.cif\n",
      "{'1': '29', '2': '28', '3': '27', '4': '26', '5': '25', '6': '24', '7': '23', '8': '22', '9': '21', '10': '20', '11': '19'}\n",
      "We did it\n",
      "UAGCAGCGGGCCCGUGGAAUCUGCUGUGA\n",
      "(((((((((((.......)))))))))))\n",
      "We are done\n",
      "1\n",
      "NOT CHARACTERIZED YET\n",
      "C4_7K00_a_G2304_a_U2312.cif\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing file 'C4_7K00_a_G2304_a_U2312.cif'\n",
      "[i] 'C4_7K00_a_G2304_a_U2312.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 21\n",
      "    total number of base pairs: 8\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C4_7UNR_A_G2291_A_U2299.cif'\n",
      "[i] 'C4_7UNR_A_G2291_A_U2299.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 21\n",
      "    total number of base pairs: 7\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of splayed-apart dinucleotides: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C4_7U2H_1A_G2304_1A_U2312.cif'\n",
      "[i] 'C4_7U2H_1A_G2304_1A_U2312.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 21\n",
      "    total number of base pairs: 8\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C4_6W6P_A_G2318_A_U2326.cif'\n",
      "[i] 'C4_6W6P_A_G2318_A_U2326.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 21\n",
      "    total number of base pairs: 8\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C4_7MT7_A_G2542_A_U2550.cif'\n",
      "[i] 'C4_7MT7_A_G2542_A_U2550.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 21\n",
      "    total number of base pairs: 8\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n",
      "\n",
      "Processing file 'C4_7NHN_A_G2337_A_U2345.cif'\n",
      "[i] 'C4_7NHN_A_G2337_A_U2345.cif' taken as in .cif format by file extension.\n",
      "    total number of nucleotides: 21\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': '21', '2': '20', '3': '19', '4': '18', '5': '17', '6': '16', '7': '15', '11': '14'}\n",
      "We did it\n",
      "AUCCUGGUCGGACAUCAGGAG\n",
      "(((((((...(..))))))))\n",
      "We are done\n",
      "2\n",
      "NOT CHARACTERIZED YET\n",
      "C4_7UNR_A_G2291_A_U2299.cif\n",
      "{'1': '21', '2': '20', '3': '19', '4': '18', '5': '17', '6': '16', '7': '15'}\n",
      "We did it\n",
      "AGACCGGUCGGAAAUCGGUCG\n",
      "(((((((.......)))))))\n",
      "We are done\n",
      "3\n",
      "NOT CHARACTERIZED YET\n",
      "C4_7U2H_1A_G2304_1A_U2312.cif\n",
      "{'1': '21', '2': '20', '3': '19', '4': '18', '5': '17', '6': '16', '7': '15', '11': '14'}\n",
      "We did it\n",
      "AGGCGGGACGGAAAUCCGCCG\n",
      "(((((((...(..))))))))\n",
      "We are done\n",
      "4\n",
      "NOT CHARACTERIZED YET\n",
      "C4_6W6P_A_G2318_A_U2326.cif\n",
      "{'1': '21', '2': '20', '3': '19', '4': '18', '5': '17', '6': '16', '7': '15', '11': '14'}\n",
      "We did it\n",
      "AGAAUGGUUGGAAAUCAUUCG\n",
      "(((((((...(..))))))))\n",
      "We are done\n",
      "5\n",
      "NOT CHARACTERIZED YET\n",
      "C4_7MT7_A_G2542_A_U2550.cif\n",
      "{'1': '21', '2': '20', '3': '19', '4': '18', '5': '17', '6': '16', '7': '15', '11': '14'}\n",
      "We did it\n",
      "AACCUGGACGGCAAUCAGGUG\n",
      "(((((((...(..))))))))\n",
      "We are done\n",
      "6\n",
      "NOT CHARACTERIZED YET\n",
      "C4_7NHN_A_G2337_A_U2345.cif\n",
      "{'1': '21', '2': '20', '3': '19', '4': '18', '5': '17', '6': '16', '7': '15', '11': '14'}\n",
      "We did it\n",
      "AGAAUGGAUGGAAAUCAUUCG\n",
      "(((((((...(..))))))))\n",
      "We are done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "    total number of base pairs: 8\n",
      "    total number of helices: 1\n",
      "    total number of stems: 1\n",
      "    total number of atom-base capping interactions: 1\n",
      "    total number of hairpin loops: 1\n",
      "    total number of non-loop single-stranded segments: 2\n",
      "\n",
      "Time used: 00:00:00:00\n"
     ]
    }
   ],
   "source": [
    "chr_dbn_fa(C1_dir, 3)\n",
    "chr_dbn_fa(C2_dir, 4)\n",
    "chr_dbn_fa(C3_dir, 9)\n",
    "chr_dbn_fa(C4_dir, 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# multiple sequence alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#here we will visualize the multiple sequence alignment for the flanking sequences of shifted wobbles in a cluster\n",
    "#the consensus sequence assigned by pymsaviz is not correct, we will apply Cavener rule later\n",
    "#we will also change the color for G and U assosiated in shifted wobble using Adobe Illustrator\n",
    "from pymsaviz import MsaViz\n",
    "\n",
    "def draw_msa(wdir):\n",
    "    dir= os.getcwd()\n",
    "    os.chdir(wdir)\n",
    "    for filename in os.listdir('.'):\n",
    "        if filename.endswith('.fa'):\n",
    "            fname= filename.rstrip('.fa')\n",
    "            mv = MsaViz(filename, wrap_length=60, show_count= False, show_grid=True, show_consensus=True, sort= False, color_scheme= 'Nucleotide')\n",
    "            mv.savefig(fname+\".pdf\")\n",
    "        else:\n",
    "            pass\n",
    "    os.chdir(dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "draw_msa(C1_dir)\n",
    "draw_msa(C2_dir)\n",
    "draw_msa(C3_dir)\n",
    "draw_msa(C4_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# visualizing secondary structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this function will read a fasta file of multiple sequences to assign a consensus sequence \n",
    "#for this the function will apply Cavener rule \n",
    "def assign_consensus_seq(dir_fa):\n",
    "    #opening the fasta file to create the list of sequences \n",
    "    cl_fa= open(dir_fa)\n",
    "    sequences=[]\n",
    "    for line in cl_fa:\n",
    "        if line.startswith('>'):\n",
    "            pass\n",
    "        else:\n",
    "            sequences.append(line.strip())\n",
    "    #IUPAC code for consensus sequence\n",
    "    con_code= {'R': ['A', 'G'],\n",
    "               'Y': ['C', 'U'],\n",
    "               'S': ['C', 'G'],\n",
    "               'W': ['A', 'U'],\n",
    "               'K': ['G', 'U'],\n",
    "               'M': ['A', 'C'],\n",
    "               'B': ['C', 'G', 'U'],\n",
    "               'D': ['A', 'G', 'U'],\n",
    "               'H': ['A', 'C', 'U'],\n",
    "               'V': ['A', 'C', 'G'],\n",
    "               'N': ['A', 'C', 'G', 'U']}\n",
    "    \n",
    "    # Convert list of sequences to a DataFrame (each column is a position)\n",
    "    df = pd.DataFrame([list(seq) for seq in sequences])\n",
    "    print (df)\n",
    "    consensus_sequence = []\n",
    "    \n",
    "    # Iterate over each column (position in the alignment)\n",
    "    count =0\n",
    "    for col in df:\n",
    "        count +=1\n",
    "        print ('Working in position ------------> '+ str(count))\n",
    "        #print ('working here '+ str(col))\n",
    "        nucleotides = df[col]\n",
    "        print (nucleotides)\n",
    "\n",
    "        # Count the occurrences of each unique item in the column\n",
    "        counts = df[col].value_counts()\n",
    "\n",
    "        # Calculate the percentage of each item and store it in a dictionary\n",
    "        nucleotide_counts = (counts / len(df) * 100).to_dict()\n",
    "\n",
    "        #sort the dictionary in descending order of the frequency of each item\n",
    "        sorted_nucleotides = dict(sorted(nucleotide_counts.items(), key=lambda item: item[1], reverse=True))\n",
    "        print (sorted_nucleotides)\n",
    "        \n",
    "        #applying Cavener rule to assign consensus sequence\n",
    "        #if only one unique item, them assign that nt as the consensus seq of that position\n",
    "        if len(list(sorted_nucleotides.keys()))==1:\n",
    "            consensus_sequence.append(list(sorted_nucleotides.keys())[0])\n",
    "        #if there are more than one unique item \n",
    "        else:\n",
    "            if sorted_nucleotides[list(sorted_nucleotides.keys())[0]]> 2*sorted_nucleotides[list(sorted_nucleotides.keys())[1]]:\n",
    "                consensus_sequence.append(list(sorted_nucleotides.keys())[0])\n",
    "            \n",
    "            elif len(list(sorted_nucleotides.keys()))>2:\n",
    "                if sorted_nucleotides[list(sorted_nucleotides.keys())[0]]+ sorted_nucleotides[list(sorted_nucleotides.keys())[1]]> 75:\n",
    "                    most_freqs= [list(sorted_nucleotides.keys())[0], list(sorted_nucleotides.keys())[1]]\n",
    "                    most_freqs.sort()\n",
    "                    for key in con_code:\n",
    "                        if most_freqs == con_code[key]:\n",
    "                            consensus_sequence.append(key)\n",
    "                else:\n",
    "                    consensus_sequence.append('N')\n",
    "            \n",
    "            elif len(list(sorted_nucleotides.keys()))==2: \n",
    "                most_freqs= [list(sorted_nucleotides.keys())[0], list(sorted_nucleotides.keys())[1]]\n",
    "                most_freqs.sort()\n",
    "                #print (most_freqs)\n",
    "                for key in con_code:\n",
    "                    #print (con_code[key])\n",
    "                    if most_freqs == con_code[key]:\n",
    "                        consensus_sequence.append(key)\n",
    "            else:\n",
    "                consensus_sequence.append('N')\n",
    "            \n",
    "        print (consensus_sequence)\n",
    "    \n",
    "    return ''.join(consensus_sequence)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this function will assign a concensus secondary structure of a cluster\n",
    "#for this the function will consider the structure with the highest frequency\n",
    "def assign_consensus_ss(cdir):\n",
    "    home= os.getcwd()\n",
    "    os.chdir(cdir)\n",
    "    dbns={}\n",
    "    for filename in os.listdir('.'):\n",
    "        if filename.endswith('.dbn'):\n",
    "            df= pd.read_csv(filename)\n",
    "            col= df.columns[0]\n",
    "            dbn= (df[col].to_list()[1])\n",
    "            if dbn in dbns:\n",
    "                dbns[dbn]+=1\n",
    "            else:\n",
    "                dbns[dbn] =1\n",
    "    print (dbns)            \n",
    "    cons_sd = max(dbns, key=dbns.get)\n",
    "    #assignin '(' and ')' for shifted wobble \n",
    "    #cons_sd_n= cons_sd[:3]+'('+cons_sd[4: len(cons_sd)-4]+ ')'+ cons_sd[len(cons_sd)-3:]\n",
    "    #print (cons_sd_n)\n",
    "    os.chdir(home)\n",
    "    return cons_sd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16 17 18 19\n",
      "0  U  G  U  U  G  G  G  U  U  A  A  G  U  C  C  C  G  C  A  A\n",
      "1  U  G  U  U  G  G  G  U  U  A  A  G  U  C  C  C  G  C  A  A\n",
      "2  U  G  U  U  G  G  G  U  U  A  A  G  U  C  C  C  G  U  A  A\n",
      "3  U  G  U  U  G  G  G  U  U  A  A  G  U  C  C  C  G  C  A  A\n",
      "4  U  G  U  U  G  G  G  U  U  A  A  G  U  C  C  C  G  C  A  A\n",
      "Working in position ------------> 1\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "Name: 0, dtype: object\n",
      "{'U': 100.0}\n",
      "['U']\n",
      "Working in position ------------> 2\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "Name: 1, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'G']\n",
      "Working in position ------------> 3\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "Name: 2, dtype: object\n",
      "{'U': 100.0}\n",
      "['U', 'G', 'U']\n",
      "Working in position ------------> 4\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "Name: 3, dtype: object\n",
      "{'U': 100.0}\n",
      "['U', 'G', 'U', 'U']\n",
      "Working in position ------------> 5\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "Name: 4, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G']\n",
      "Working in position ------------> 6\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "Name: 5, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G']\n",
      "Working in position ------------> 7\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "Name: 6, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G']\n",
      "Working in position ------------> 8\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "Name: 7, dtype: object\n",
      "{'U': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U']\n",
      "Working in position ------------> 9\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "Name: 8, dtype: object\n",
      "{'U': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U']\n",
      "Working in position ------------> 10\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "Name: 9, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A']\n",
      "Working in position ------------> 11\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "Name: 10, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A']\n",
      "Working in position ------------> 12\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "Name: 11, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G']\n",
      "Working in position ------------> 13\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "Name: 12, dtype: object\n",
      "{'U': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G', 'U']\n",
      "Working in position ------------> 14\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "Name: 13, dtype: object\n",
      "{'C': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G', 'U', 'C']\n",
      "Working in position ------------> 15\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "Name: 14, dtype: object\n",
      "{'C': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G', 'U', 'C', 'C']\n",
      "Working in position ------------> 16\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "Name: 15, dtype: object\n",
      "{'C': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G', 'U', 'C', 'C', 'C']\n",
      "Working in position ------------> 17\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "Name: 16, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G', 'U', 'C', 'C', 'C', 'G']\n",
      "Working in position ------------> 18\n",
      "0    C\n",
      "1    C\n",
      "2    U\n",
      "3    C\n",
      "4    C\n",
      "Name: 17, dtype: object\n",
      "{'C': 80.0, 'U': 20.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G', 'U', 'C', 'C', 'C', 'G', 'C']\n",
      "Working in position ------------> 19\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "Name: 18, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G', 'U', 'C', 'C', 'C', 'G', 'C', 'A']\n",
      "Working in position ------------> 20\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "Name: 19, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'G', 'U', 'U', 'G', 'G', 'G', 'U', 'U', 'A', 'A', 'G', 'U', 'C', 'C', 'C', 'G', 'C', 'A', 'A']\n",
      "{'.(.((((((.).)))))..)': 3, '.(.(((((....)))))..)': 2}\n"
     ]
    }
   ],
   "source": [
    "c1_con_seq= assign_consensus_seq('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/clusters_with_finalized_flanks/C1/C1_seqs.fa')\n",
    "c1_con_ss= assign_consensus_ss('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/clusters_with_finalized_flanks/C1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0  1  2  3  4  5  6  7  8  9   ... 35 36 37 38 39 40 41 42 43 44\n",
      "0  G  G  A  A  U  U  U  C  C  U  ...  A  G  G  A  A  G  G  A  A  C\n",
      "1  G  G  A  A  U  U  C  C  U  G  ...  C  A  G  G  A  G  G  A  A  C\n",
      "2  A  G  A  A  U  U  C  C  A  G  ...  C  U  G  G  A  G  G  A  A  U\n",
      "3  G  G  A  A  U  U  C  C  C  G  ...  C  G  G  G  A  G  G  A  A  C\n",
      "4  A  G  A  A  U  U  C  C  A  G  ...  C  U  G  G  A  G  G  A  A  U\n",
      "5  G  G  A  A  U  U  C  C  A  C  ...  G  U  G  G  A  G  G  A  A  C\n",
      "6  G  G  A  A  U  U  C  C  A  U  ...  A  U  G  G  A  G  G  A  A  C\n",
      "7  G  G  A  A  U  U  C  C  A  C  ...  G  U  G  G  A  G  G  A  A  C\n",
      "8  U  G  G  A  U  A  G  U  U  G  ...  U  G  A  C  U  G  G  A  C  G\n",
      "\n",
      "[9 rows x 45 columns]\n",
      "Working in position ------------> 1\n",
      "0    G\n",
      "1    G\n",
      "2    A\n",
      "3    G\n",
      "4    A\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    U\n",
      "Name: 0, dtype: object\n",
      "{'G': 66.66666666666666, 'A': 22.22222222222222, 'U': 11.11111111111111}\n",
      "['G']\n",
      "Working in position ------------> 2\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 1, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G']\n",
      "Working in position ------------> 3\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    G\n",
      "Name: 2, dtype: object\n",
      "{'A': 88.88888888888889, 'G': 11.11111111111111}\n",
      "['G', 'G', 'A']\n",
      "Working in position ------------> 4\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    A\n",
      "Name: 3, dtype: object\n",
      "{'A': 100.0}\n",
      "['G', 'G', 'A', 'A']\n",
      "Working in position ------------> 5\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "5    U\n",
      "6    U\n",
      "7    U\n",
      "8    U\n",
      "Name: 4, dtype: object\n",
      "{'U': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U']\n",
      "Working in position ------------> 6\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "5    U\n",
      "6    U\n",
      "7    U\n",
      "8    A\n",
      "Name: 5, dtype: object\n",
      "{'U': 88.88888888888889, 'A': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U']\n",
      "Working in position ------------> 7\n",
      "0    U\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "5    C\n",
      "6    C\n",
      "7    C\n",
      "8    G\n",
      "Name: 6, dtype: object\n",
      "{'C': 77.77777777777779, 'U': 11.11111111111111, 'G': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C']\n",
      "Working in position ------------> 8\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "5    C\n",
      "6    C\n",
      "7    C\n",
      "8    U\n",
      "Name: 7, dtype: object\n",
      "{'C': 88.88888888888889, 'U': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C']\n",
      "Working in position ------------> 9\n",
      "0    C\n",
      "1    U\n",
      "2    A\n",
      "3    C\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    U\n",
      "Name: 8, dtype: object\n",
      "{'A': 55.55555555555556, 'C': 22.22222222222222, 'U': 22.22222222222222}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A']\n",
      "Working in position ------------> 10\n",
      "0    U\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    C\n",
      "6    U\n",
      "7    C\n",
      "8    G\n",
      "Name: 9, dtype: object\n",
      "{'G': 55.55555555555556, 'U': 22.22222222222222, 'C': 22.22222222222222}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G']\n",
      "Working in position ------------> 11\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 10, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G']\n",
      "Working in position ------------> 12\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    A\n",
      "4    U\n",
      "5    U\n",
      "6    U\n",
      "7    U\n",
      "8    G\n",
      "Name: 11, dtype: object\n",
      "{'U': 77.77777777777779, 'A': 11.11111111111111, 'G': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U']\n",
      "Working in position ------------> 13\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    C\n",
      "Name: 12, dtype: object\n",
      "{'G': 88.88888888888889, 'C': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G']\n",
      "Working in position ------------> 14\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "5    U\n",
      "6    U\n",
      "7    U\n",
      "8    G\n",
      "Name: 13, dtype: object\n",
      "{'U': 88.88888888888889, 'G': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U']\n",
      "Working in position ------------> 15\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    A\n",
      "Name: 14, dtype: object\n",
      "{'A': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A']\n",
      "Working in position ------------> 16\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 15, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G']\n",
      "Working in position ------------> 17\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "5    C\n",
      "6    C\n",
      "7    C\n",
      "8    A\n",
      "Name: 16, dtype: object\n",
      "{'C': 88.88888888888889, 'A': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C']\n",
      "Working in position ------------> 18\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 17, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G']\n",
      "Working in position ------------> 19\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 18, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G']\n",
      "Working in position ------------> 20\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "5    U\n",
      "6    U\n",
      "7    U\n",
      "8    U\n",
      "Name: 19, dtype: object\n",
      "{'U': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U']\n",
      "Working in position ------------> 21\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 20, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G']\n",
      "Working in position ------------> 22\n",
      "0    A\n",
      "1    G\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    A\n",
      "Name: 21, dtype: object\n",
      "{'A': 88.88888888888889, 'G': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A']\n",
      "Working in position ------------> 23\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    A\n",
      "Name: 22, dtype: object\n",
      "{'A': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A']\n",
      "Working in position ------------> 24\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    A\n",
      "Name: 23, dtype: object\n",
      "{'A': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A']\n",
      "Working in position ------------> 25\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "5    U\n",
      "6    U\n",
      "7    U\n",
      "8    U\n",
      "Name: 24, dtype: object\n",
      "{'U': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U']\n",
      "Working in position ------------> 26\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 25, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G']\n",
      "Working in position ------------> 27\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "5    C\n",
      "6    C\n",
      "7    C\n",
      "8    C\n",
      "Name: 26, dtype: object\n",
      "{'C': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C']\n",
      "Working in position ------------> 28\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 27, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G']\n",
      "Working in position ------------> 29\n",
      "0    U\n",
      "1    C\n",
      "2    U\n",
      "3    C\n",
      "4    U\n",
      "5    U\n",
      "6    C\n",
      "7    U\n",
      "8    A\n",
      "Name: 28, dtype: object\n",
      "{'U': 55.55555555555556, 'C': 33.33333333333333, 'A': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y']\n",
      "Working in position ------------> 30\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    A\n",
      "Name: 29, dtype: object\n",
      "{'A': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A']\n",
      "Working in position ------------> 31\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 30, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G']\n",
      "Working in position ------------> 32\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    A\n",
      "Name: 31, dtype: object\n",
      "{'A': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A']\n",
      "Working in position ------------> 33\n",
      "0    U\n",
      "1    U\n",
      "2    G\n",
      "3    U\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    U\n",
      "8    C\n",
      "Name: 32, dtype: object\n",
      "{'U': 44.44444444444444, 'G': 44.44444444444444, 'C': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K']\n",
      "Working in position ------------> 34\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    C\n",
      "Name: 33, dtype: object\n",
      "{'A': 88.88888888888889, 'C': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A']\n",
      "Working in position ------------> 35\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    C\n",
      "4    U\n",
      "5    U\n",
      "6    U\n",
      "7    U\n",
      "8    C\n",
      "Name: 34, dtype: object\n",
      "{'U': 77.77777777777779, 'C': 22.22222222222222}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U']\n",
      "Working in position ------------> 36\n",
      "0    A\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "5    G\n",
      "6    A\n",
      "7    G\n",
      "8    U\n",
      "Name: 35, dtype: object\n",
      "{'C': 44.44444444444444, 'A': 22.22222222222222, 'G': 22.22222222222222, 'U': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N']\n",
      "Working in position ------------> 37\n",
      "0    G\n",
      "1    A\n",
      "2    U\n",
      "3    G\n",
      "4    U\n",
      "5    U\n",
      "6    U\n",
      "7    U\n",
      "8    G\n",
      "Name: 36, dtype: object\n",
      "{'U': 55.55555555555556, 'G': 33.33333333333333, 'A': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K']\n",
      "Working in position ------------> 38\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    A\n",
      "Name: 37, dtype: object\n",
      "{'G': 88.88888888888889, 'A': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K', 'G']\n",
      "Working in position ------------> 39\n",
      "0    A\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    C\n",
      "Name: 38, dtype: object\n",
      "{'G': 77.77777777777779, 'A': 11.11111111111111, 'C': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K', 'G', 'G']\n",
      "Working in position ------------> 40\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    U\n",
      "Name: 39, dtype: object\n",
      "{'A': 88.88888888888889, 'U': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K', 'G', 'G', 'A']\n",
      "Working in position ------------> 41\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 40, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K', 'G', 'G', 'A', 'G']\n",
      "Working in position ------------> 42\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "6    G\n",
      "7    G\n",
      "8    G\n",
      "Name: 41, dtype: object\n",
      "{'G': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K', 'G', 'G', 'A', 'G', 'G']\n",
      "Working in position ------------> 43\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    A\n",
      "Name: 42, dtype: object\n",
      "{'A': 100.0}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K', 'G', 'G', 'A', 'G', 'G', 'A']\n",
      "Working in position ------------> 44\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "6    A\n",
      "7    A\n",
      "8    C\n",
      "Name: 43, dtype: object\n",
      "{'A': 88.88888888888889, 'C': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K', 'G', 'G', 'A', 'G', 'G', 'A', 'A']\n",
      "Working in position ------------> 45\n",
      "0    C\n",
      "1    C\n",
      "2    U\n",
      "3    C\n",
      "4    U\n",
      "5    C\n",
      "6    C\n",
      "7    C\n",
      "8    G\n",
      "Name: 44, dtype: object\n",
      "{'C': 66.66666666666666, 'U': 22.22222222222222, 'G': 11.11111111111111}\n",
      "['G', 'G', 'A', 'A', 'U', 'U', 'C', 'C', 'A', 'G', 'G', 'U', 'G', 'U', 'A', 'G', 'C', 'G', 'G', 'U', 'G', 'A', 'A', 'A', 'U', 'G', 'C', 'G', 'Y', 'A', 'G', 'A', 'K', 'A', 'U', 'N', 'K', 'G', 'G', 'A', 'G', 'G', 'A', 'A', 'C']\n",
      "{'(((((((((((((((((((....)))))...))))))))))))))': 4, '((((((((((((((((((((..)))))...)))))))))))))))': 1, '(((((((((((((((((((....))))...)))))))))))))))': 3, '((((((((((((((((((......))))...))))))))))))))': 1}\n"
     ]
    }
   ],
   "source": [
    "c2_con_seq= assign_consensus_seq('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/clusters_with_finalized_flanks/C2/C2_seqs.fa')\n",
    "c2_con_ss= assign_consensus_ss('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/clusters_with_finalized_flanks/C2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    }
   ],
   "source": [
    "print (len('))))))))'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0  1  2  3  4  5  6  7  8  9   ... 19 20 21 22 23 24 25 26 27 28\n",
      "0  U  A  G  C  A  G  C  G  G  G  ...  U  C  C  G  C  U  G  U  G  A\n",
      "1  U  A  G  C  A  G  C  G  G  G  ...  U  C  U  G  C  U  G  U  G  A\n",
      "2  U  A  G  C  A  U  G  G  G  G  ...  U  C  C  C  G  U  G  U  G  A\n",
      "3  U  A  C  G  A  C  G  G  A  G  ...  U  U  C  C  G  U  C  G  G  A\n",
      "\n",
      "[4 rows x 29 columns]\n",
      "Working in position ------------> 1\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "Name: 0, dtype: object\n",
      "{'U': 100.0}\n",
      "['U']\n",
      "Working in position ------------> 2\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "Name: 1, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'A']\n",
      "Working in position ------------> 3\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    C\n",
      "Name: 2, dtype: object\n",
      "{'G': 75.0, 'C': 25.0}\n",
      "['U', 'A', 'G']\n",
      "Working in position ------------> 4\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    G\n",
      "Name: 3, dtype: object\n",
      "{'C': 75.0, 'G': 25.0}\n",
      "['U', 'A', 'G', 'C']\n",
      "Working in position ------------> 5\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "Name: 4, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A']\n",
      "Working in position ------------> 6\n",
      "0    G\n",
      "1    G\n",
      "2    U\n",
      "3    C\n",
      "Name: 5, dtype: object\n",
      "{'G': 50.0, 'U': 25.0, 'C': 25.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N']\n",
      "Working in position ------------> 7\n",
      "0    C\n",
      "1    C\n",
      "2    G\n",
      "3    G\n",
      "Name: 6, dtype: object\n",
      "{'C': 50.0, 'G': 50.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S']\n",
      "Working in position ------------> 8\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "Name: 7, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G']\n",
      "Working in position ------------> 9\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    A\n",
      "Name: 8, dtype: object\n",
      "{'G': 75.0, 'A': 25.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G']\n",
      "Working in position ------------> 10\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "Name: 9, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G']\n",
      "Working in position ------------> 11\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "Name: 10, dtype: object\n",
      "{'C': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C']\n",
      "Working in position ------------> 12\n",
      "0    C\n",
      "1    C\n",
      "2    A\n",
      "3    A\n",
      "Name: 11, dtype: object\n",
      "{'C': 50.0, 'A': 50.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M']\n",
      "Working in position ------------> 13\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "Name: 12, dtype: object\n",
      "{'C': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C']\n",
      "Working in position ------------> 14\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "Name: 13, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G']\n",
      "Working in position ------------> 15\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "Name: 14, dtype: object\n",
      "{'U': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U']\n",
      "Working in position ------------> 16\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "Name: 15, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G']\n",
      "Working in position ------------> 17\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    A\n",
      "Name: 16, dtype: object\n",
      "{'G': 75.0, 'A': 25.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G']\n",
      "Working in position ------------> 18\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "Name: 17, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A']\n",
      "Working in position ------------> 19\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "Name: 18, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A']\n",
      "Working in position ------------> 20\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "Name: 19, dtype: object\n",
      "{'U': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U']\n",
      "Working in position ------------> 21\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    U\n",
      "Name: 20, dtype: object\n",
      "{'C': 75.0, 'U': 25.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C']\n",
      "Working in position ------------> 22\n",
      "0    C\n",
      "1    U\n",
      "2    C\n",
      "3    C\n",
      "Name: 21, dtype: object\n",
      "{'C': 75.0, 'U': 25.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C', 'C']\n",
      "Working in position ------------> 23\n",
      "0    G\n",
      "1    G\n",
      "2    C\n",
      "3    C\n",
      "Name: 22, dtype: object\n",
      "{'G': 50.0, 'C': 50.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C', 'C', 'S']\n",
      "Working in position ------------> 24\n",
      "0    C\n",
      "1    C\n",
      "2    G\n",
      "3    G\n",
      "Name: 23, dtype: object\n",
      "{'C': 50.0, 'G': 50.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C', 'C', 'S', 'S']\n",
      "Working in position ------------> 25\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "Name: 24, dtype: object\n",
      "{'U': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C', 'C', 'S', 'S', 'U']\n",
      "Working in position ------------> 26\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    C\n",
      "Name: 25, dtype: object\n",
      "{'G': 75.0, 'C': 25.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C', 'C', 'S', 'S', 'U', 'G']\n",
      "Working in position ------------> 27\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    G\n",
      "Name: 26, dtype: object\n",
      "{'U': 75.0, 'G': 25.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C', 'C', 'S', 'S', 'U', 'G', 'U']\n",
      "Working in position ------------> 28\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "Name: 27, dtype: object\n",
      "{'G': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C', 'C', 'S', 'S', 'U', 'G', 'U', 'G']\n",
      "Working in position ------------> 29\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "Name: 28, dtype: object\n",
      "{'A': 100.0}\n",
      "['U', 'A', 'G', 'C', 'A', 'N', 'S', 'G', 'G', 'G', 'C', 'M', 'C', 'G', 'U', 'G', 'G', 'A', 'A', 'U', 'C', 'C', 'S', 'S', 'U', 'G', 'U', 'G', 'A']\n",
      "{'(((((((((((.......)))))))))))': 3, '(((((((((((...)..).).))))))))': 1}\n"
     ]
    }
   ],
   "source": [
    "c3_con_seq= assign_consensus_seq('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/clusters_with_finalized_flanks/C3/C3_seqs.fa')\n",
    "c3_con_ss= assign_consensus_ss('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/clusters_with_finalized_flanks/C3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0  1  2  3  4  5  6  7  8  9   ... 11 12 13 14 15 16 17 18 19 20\n",
      "0  A  G  A  C  C  G  G  U  C  G  ...  A  A  A  U  C  G  G  U  C  G\n",
      "1  A  U  C  C  U  G  G  U  C  G  ...  A  C  A  U  C  A  G  G  A  G\n",
      "2  A  G  A  A  U  G  G  U  U  G  ...  A  A  A  U  C  A  U  U  C  G\n",
      "3  A  A  C  C  U  G  G  A  C  G  ...  C  A  A  U  C  A  G  G  U  G\n",
      "4  A  G  G  C  G  G  G  A  C  G  ...  A  A  A  U  C  C  G  C  C  G\n",
      "5  A  G  A  A  U  G  G  A  U  G  ...  A  A  A  U  C  A  U  U  C  G\n",
      "\n",
      "[6 rows x 21 columns]\n",
      "Working in position ------------> 1\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "Name: 0, dtype: object\n",
      "{'A': 100.0}\n",
      "['A']\n",
      "Working in position ------------> 2\n",
      "0    G\n",
      "1    U\n",
      "2    G\n",
      "3    A\n",
      "4    G\n",
      "5    G\n",
      "Name: 1, dtype: object\n",
      "{'G': 66.66666666666666, 'U': 16.666666666666664, 'A': 16.666666666666664}\n",
      "['A', 'G']\n",
      "Working in position ------------> 3\n",
      "0    A\n",
      "1    C\n",
      "2    A\n",
      "3    C\n",
      "4    G\n",
      "5    A\n",
      "Name: 2, dtype: object\n",
      "{'A': 50.0, 'C': 33.33333333333333, 'G': 16.666666666666664}\n",
      "['A', 'G', 'M']\n",
      "Working in position ------------> 4\n",
      "0    C\n",
      "1    C\n",
      "2    A\n",
      "3    C\n",
      "4    C\n",
      "5    A\n",
      "Name: 3, dtype: object\n",
      "{'C': 66.66666666666666, 'A': 33.33333333333333}\n",
      "['A', 'G', 'M', 'M']\n",
      "Working in position ------------> 5\n",
      "0    C\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    G\n",
      "5    U\n",
      "Name: 4, dtype: object\n",
      "{'U': 66.66666666666666, 'C': 16.666666666666664, 'G': 16.666666666666664}\n",
      "['A', 'G', 'M', 'M', 'U']\n",
      "Working in position ------------> 6\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "Name: 5, dtype: object\n",
      "{'G': 100.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G']\n",
      "Working in position ------------> 7\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "Name: 6, dtype: object\n",
      "{'G': 100.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G']\n",
      "Working in position ------------> 8\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "Name: 7, dtype: object\n",
      "{'U': 50.0, 'A': 50.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W']\n",
      "Working in position ------------> 9\n",
      "0    C\n",
      "1    C\n",
      "2    U\n",
      "3    C\n",
      "4    C\n",
      "5    U\n",
      "Name: 8, dtype: object\n",
      "{'C': 66.66666666666666, 'U': 33.33333333333333}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y']\n",
      "Working in position ------------> 10\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "Name: 9, dtype: object\n",
      "{'G': 100.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G']\n",
      "Working in position ------------> 11\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "Name: 10, dtype: object\n",
      "{'G': 100.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G']\n",
      "Working in position ------------> 12\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    C\n",
      "4    A\n",
      "5    A\n",
      "Name: 11, dtype: object\n",
      "{'A': 83.33333333333334, 'C': 16.666666666666664}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A']\n",
      "Working in position ------------> 13\n",
      "0    A\n",
      "1    C\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "Name: 12, dtype: object\n",
      "{'A': 83.33333333333334, 'C': 16.666666666666664}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A']\n",
      "Working in position ------------> 14\n",
      "0    A\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    A\n",
      "5    A\n",
      "Name: 13, dtype: object\n",
      "{'A': 100.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A', 'A']\n",
      "Working in position ------------> 15\n",
      "0    U\n",
      "1    U\n",
      "2    U\n",
      "3    U\n",
      "4    U\n",
      "5    U\n",
      "Name: 14, dtype: object\n",
      "{'U': 100.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A', 'A', 'U']\n",
      "Working in position ------------> 16\n",
      "0    C\n",
      "1    C\n",
      "2    C\n",
      "3    C\n",
      "4    C\n",
      "5    C\n",
      "Name: 15, dtype: object\n",
      "{'C': 100.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A', 'A', 'U', 'C']\n",
      "Working in position ------------> 17\n",
      "0    G\n",
      "1    A\n",
      "2    A\n",
      "3    A\n",
      "4    C\n",
      "5    A\n",
      "Name: 16, dtype: object\n",
      "{'A': 66.66666666666666, 'G': 16.666666666666664, 'C': 16.666666666666664}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A', 'A', 'U', 'C', 'A']\n",
      "Working in position ------------> 18\n",
      "0    G\n",
      "1    G\n",
      "2    U\n",
      "3    G\n",
      "4    G\n",
      "5    U\n",
      "Name: 17, dtype: object\n",
      "{'G': 66.66666666666666, 'U': 33.33333333333333}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A', 'A', 'U', 'C', 'A', 'K']\n",
      "Working in position ------------> 19\n",
      "0    U\n",
      "1    G\n",
      "2    U\n",
      "3    G\n",
      "4    C\n",
      "5    U\n",
      "Name: 18, dtype: object\n",
      "{'U': 50.0, 'G': 33.33333333333333, 'C': 16.666666666666664}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A', 'A', 'U', 'C', 'A', 'K', 'K']\n",
      "Working in position ------------> 20\n",
      "0    C\n",
      "1    A\n",
      "2    C\n",
      "3    U\n",
      "4    C\n",
      "5    C\n",
      "Name: 19, dtype: object\n",
      "{'C': 66.66666666666666, 'A': 16.666666666666664, 'U': 16.666666666666664}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A', 'A', 'U', 'C', 'A', 'K', 'K', 'C']\n",
      "Working in position ------------> 21\n",
      "0    G\n",
      "1    G\n",
      "2    G\n",
      "3    G\n",
      "4    G\n",
      "5    G\n",
      "Name: 20, dtype: object\n",
      "{'G': 100.0}\n",
      "['A', 'G', 'M', 'M', 'U', 'G', 'G', 'W', 'Y', 'G', 'G', 'A', 'A', 'A', 'U', 'C', 'A', 'K', 'K', 'C', 'G']\n",
      "{'(((((((...(..))))))))': 5, '(((((((.......)))))))': 1}\n"
     ]
    }
   ],
   "source": [
    "c4_con_seq= assign_consensus_seq('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/clusters_with_finalized_flanks/C4/C4_seqs.fa')\n",
    "c4_con_ss= assign_consensus_ss('/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/clusters_with_finalized_flanks/C4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FOR CLUSTER 2\n",
      "GGAAUUCCAGGUGUAGCGGUGAAAUGCGYAGAKAUNKGGAGGAAC\n",
      "(((((((((((((((((((....)))))...))))))))))))))\n",
      " \n"
     ]
    }
   ],
   "source": [
    "#print ('FOR CLUSTER 1')\n",
    "#print (c1_con_seq)\n",
    "#print (c1_con_ss)\n",
    "#print (' ')\n",
    "print ('FOR CLUSTER 2')\n",
    "print (c2_con_seq)\n",
    "print (c2_con_ss)\n",
    "print (' ')\n",
    "#print ('FOR CLUSTER 3')\n",
    "#print (c3_con_seq)\n",
    "#print (c3_con_ss)\n",
    "#print (' ')\n",
    "#print ('FOR CLUSTER 4')\n",
    "#print (c4_con_seq)\n",
    "#print (c4_con_ss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAAUUCCAGG\n",
      "UGUAGCGGUG\n",
      "AAAUGCGYAG\n",
      "AKAUNKGGAG\n",
      "GAA\n"
     ]
    }
   ],
   "source": [
    "seq= 'GAAUUCCAGGUGUAGCGGUGAAAUGCGYAGAKAUNKGGAGGAA'\n",
    "\n",
    "print (seq[:10])\n",
    "print (seq[10:20])\n",
    "print (seq[20:30])\n",
    "print (seq[30:40])\n",
    "print (seq[40:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "C1_test= {'1':'6', '2': '5', '9': '27', '11': '24', '12':'23', '13':'22', '14':'21', '15': '20', '16': '18'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [],
   "source": [
    "dbn_test= ['.']*34\n",
    "\n",
    "for i in C1_test:\n",
    "    dbn_test[int(i)-1]= '('\n",
    "    dbn_test[int(C1_test[i])-1]= ')'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['(', '(', '.', '.', ')', ')', '.', '.', '(', '.', '(', '(', '(', '(', '(', '(', '.', ')', '.', ')', ')', ')', ')', ')', '.', '.', ')', '.', '.', '.', '.', '.', '.', '.']\n",
      "((..))..(.((((((.).)))))..).......\n"
     ]
    }
   ],
   "source": [
    "print (dbn_test)\n",
    "\n",
    "dbn_test2= ''.join(dbn_test)\n",
    "print (dbn_test2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(C2_dir)\n",
    "F= 'cluster_2_consensus_2D_new'\n",
    "whole_seq= C2_seq\n",
    "whole_ss= C2_dbn\n",
    "all_res= ','.join([str(i+1) for i, j in enumerate(whole_seq) if i!= 3 and i!= len(whole_seq)-4])\n",
    "hi_res= '4,'+ str(len(whole_ss)-3)\n",
    "def run_varna_radiate_colored(rna_sequence, rna_dot_bracket, segment1= '', segment2= '', output_file = F+'.png'):\n",
    "    #subprocess.Popen([\"java\", \"-cp\",\"./VARNAv3-93.jar\", \"fr.orsay.lri.varna.applications.VARNAcmd\", '-sequenceDBN', rna_sequence, '-structureDBN', rna_dot_bracket, '-o', output_file, '-algorithm', 'radiate', '-resolution', '8.0', '-bpStyle', 'lw', '-basesStyle1', 'fill=#B8499B,outline=#B8499B,label=#000000,number=#FF0000','-applyBasesStyle1on', segment1], stderr=subprocess.STDOUT, stdout=subprocess.PIPE).communicate()[0]\n",
    "    #subprocess.Popen([\"java\", \"-cp\",\"./VARNAv3-93.jar\", \"fr.orsay.lri.varna.applications.VARNAcmd\", '-sequenceDBN', rna_sequence, '-structureDBN', rna_dot_bracket, '-o', output_file, '-algorithm', 'radiate', '-resolution', '8.0', '-bpStyle', 'line', '-basesStyle1', 'fill=#B8499B,outline=#B8499B,label=#000000,number=#FF0000','-applyBasesStyle1on', segment1], stderr=subprocess.STDOUT, stdout=subprocess.PIPE).communicate()[0]\n",
    "    subprocess.Popen([\"java\", \"-cp\",\"./VARNAv3-93.jar\", \"fr.orsay.lri.varna.applications.VARNAcmd\", \"-sequenceDBN\", rna_sequence, \"-structureDBN\", rna_dot_bracket, \"-o\", output_file, \"-algorithm\", \"linear\", \"-resolution\", \"10.0\", \"-basesStyle1\", \"fill=#B8499B,outline=#B8499B,label=#000000,number=#FF0000\",\"-applyBasesStyle1on\", segment1, \"-basesStyle2\", \"fill=#FFFFFF,outline=#000000,label=#000000,number=#000000\",\"-applyBasesStyle2on\", segment2, \"-bpStyle\", \"Line\" ], stderr=subprocess.STDOUT, stdout=subprocess.PIPE).communicate()[0]\n",
    "    #subprocess.run([\"java\", \"-cp\",\"./VARNAv3-93.jar\", \"fr.orsay.lri.varna.applications.VARNAcmd\", '-sequenceDBN', rna_sequence, '-structureDBN', rna_dot_bracket, '-o', output_file, '-algorithm', 'linear', '-resolution', '8.0', '-bpStyle', 'line', '-basesStyle1', 'fill=#B8499B,outline=#B8499B,label=#000000,number=#FF0000','-applyBasesStyle1on', segment1, '-basesStyle2', 'fill=#FFFFFF,outline=#000000,label=#000000,number=#000000','-applyBasesStyle2on', segment2], check= True)\n",
    "\n",
    "run_varna_radiate_colored(whole_seq, whole_ss, segment1= hi_res, segment2= all_res, output_file = F+'.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sharear/Documents/sky/RESEARCH_&_GRAD_SCHOOL/Penn_State/anionic_G_U/get_all_RNA/new_data_April_2023/2nd_version_algorithm_removing_redundancy/anionic_GU_pipeline/anionic_GU_pipeline_updated/identifying_shifted_wobbles/analysis/august_2024_clusters/C2\n"
     ]
    }
   ],
   "source": [
    "print (os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
